motionplan/utils.go

```diff
@@ -98,3 +45,245 @@ func (r *resultPromise) result() ([]node, error) {
+	inputs := map[string][]referenceframe.Input{}
+	i := 0
+	for _, frame := range lfs.frames {
+		if len(frame.DoF()) == 0 {
+			continue
+		}
+		frameInputs := make([]referenceframe.Input, len(frame.DoF()))
+		for j := range frame.DoF() {
+			if i >= len(floatSlice) {
+				return nil, fmt.Errorf("not enough values in float slice for frame %s", frame.Name())
+			}
+			frameInputs[j] = referenceframe.Input{Value: floatSlice[i]}
+			i++
+		}
+		inputs[frame.Name()] = frameInputs
+	}
+	return inputs, nil
+}
+
+// motionChain structs are meant to be ephemerally created for each individual goal in a motion request, and calculates the shortest
+// path between components in the framesystem allowing knowledge of which frames may move.
+type motionChain struct {
+	// List of names of all frames that could move, used for collision detection
+	// As an example a gripper attached to an arm which is moving relative to World, would not be in frames below but in this object
+	movingFS       referenceframe.FrameSystem
+	frames         []referenceframe.Frame // all frames directly between and including solveFrame and goalFrame. Order not important.
+	solveFrameName string
+	goalFrameName  string
+	// If this is true, then goals are translated to their position in `World` before solving.
+	// This is useful when e.g. moving a gripper relative to a point seen by a camera built into that gripper
+	// TODO(pl): explore allowing this to be frames other than world
+	worldRooted bool
+}
+
+func motionChainFromGoal(fs referenceframe.FrameSystem, moveFrame string, goal *referenceframe.PoseInFrame) (*motionChain, error) {
+	// get goal frame
+	goalFrame := fs.Frame(goal.Parent())
+	if goalFrame == nil {
+		return nil, referenceframe.NewFrameMissingError(goal.Parent())
+	}
+	goalFrameList, err := fs.TracebackFrame(goalFrame)
+	if err != nil {
+		return nil, err
+	}
+
+	// get solve frame
+	solveFrame := fs.Frame(moveFrame)
+	if solveFrame == nil {
+		return nil, referenceframe.NewFrameMissingError(moveFrame)
+	}
+	solveFrameList, err := fs.TracebackFrame(solveFrame)
+	if err != nil {
+		return nil, err
+	}
+	if len(solveFrameList) == 0 {
+		return nil, errors.New("solveFrameList was empty")
+	}
+
+	movingFS := func(frameList []referenceframe.Frame) (referenceframe.FrameSystem, error) {
+		// Find first moving frame
+		var moveF referenceframe.Frame
+		for i := len(frameList) - 1; i >= 0; i-- {
+			if len(frameList[i].DoF()) != 0 {
+				moveF = frameList[i]
+				break
+			}
+		}
+		if moveF == nil {
+			return referenceframe.NewEmptyFrameSystem(""), nil
+		}
+		return fs.FrameSystemSubset(moveF)
+	}
+
+	// find pivot frame between goal and solve frames
+	var moving referenceframe.FrameSystem
+	var frames []referenceframe.Frame
+	worldRooted := false
+	pivotFrame, err := findPivotFrame(solveFrameList, goalFrameList)
+	if err != nil {
+		return nil, err
+	}
+	if pivotFrame.Name() == referenceframe.World {
+		frames = uniqInPlaceSlice(append(solveFrameList, goalFrameList...))
```

Comment:
Can't you just query the FrameNames of the merged FS instead of having this function?

Reply:
No, because the framesystem may have more than two branches at the pivot point.